feat(core): implement exponential polling strategy

archerzz opened the pull request:
Implements results of https://github.com/Azure/azure-sdk/issues/3841

1. add `IOperationPollingStrategy` and 2 implementation
  1.1 `ConstantPollingStrategy` is for constant polling interval
  1.2 `ExponentialPollingStrategy` is for exponential polling interval starting from 1 second up to 32 seconds
2. refactor `OperationInternalBase` and `OperationInternalOfT`:
  2.1 replace `DefaultPollingInterval` with `PollingStrategy`
  2.2 override implementation of `WaitForCompletionResponseAsync` and `WaitForCompletionAsync`, use interval from `PollingStrategy.PollingInterval`
3. add unit test cases, and update project build dependencies

part of #26922

# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

archerzz committed:
feat(core): implement exponential polling strategy

1. add `IOperationPollingStrategy` and 2 implementation
  1.1 `ConstantPollingStrategy` is for constant polling interval
  1.2 `ExponentialPollingStrategy` is for exponential polling interval starting from 1 second up to 32 seconds
2. refactor `OperationInternalBase` and `OperationInternalOfT`:
  2.1 replace `DefaultPollingInterval` with `PollingStrategy`
  2.2 override implementation of `WaitForCompletionResponseAsync` and `WaitForCompletionAsync`, use interval from `PollingStrategy.PollingInterval`
3. add unit test cases, and update project build dependencies

part of #26922

christothes commented:
I'm hesitant to introduce another delay strategy considering recent discussions about the intent to create a new centralized Delay concept (see #26851). Is there a way we could satisfy the immediate need of this PR without introducing these new polling strategy types?

m-nash commented on sdk/core/Azure.Core/src/Shared/OperationInternalBase.cs:
```diff
@@ -53,10 +53,15 @@ protected OperationInternalBase(ClientDiagnostics clientDiagnostics, Response ra
+        /// <summary>
+        /// Gets the polling interval.
+        /// </summary>
+        protected TimeSpan PollingInterval { get => PollingStrategy.PollingInterval; }
```
this should be a member of the constant polling strategy class

m-nash commented on sdk/core/Azure.Core/src/Shared/OperationInternalBase.cs:
```diff
@@ -27,7 +27,7 @@ protected OperationInternalBase(ClientDiagnostics clientDiagnostics, Response ra
             _scopeAttributes = scopeAttributes?.ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
             RawResponse = rawResponse;
-            DefaultPollingInterval = TimeSpan.FromSeconds(1);
+            PollingStrategy = new ConstantPollingStrategy();
```
We should check if Retry-After is in the rawResponse if so instantiate a RetryAfterStrategy.  If there is ever a case where some of the responses will have Retry-After but others won't then RetryAfterStrategy can take in a fallback, but this shouldn't be needed.

m-nash commented on sdk/core/Azure.Core/src/Shared/OperationInternalBase.cs:
```diff
@@ -114,8 +119,21 @@ protected OperationInternalBase(ClientDiagnostics clientDiagnostics, Response ra
+                }
+
+                TimeSpan delay = GetServerDelay(response, PollingInterval);
+                await WaitAsync(delay, cancellationToken).ConfigureAwait(false);
```
This should just ask the polling strategy to give you next wait time
```suggestion
                await WaitAsync(PollingStrategy.GetNextWait(), cancellationToken).ConfigureAwait(false);
```

m-nash commented on sdk/core/Azure.Core/src/Shared/OperationInternalBase.cs:
```diff
@@ -114,8 +119,21 @@ protected OperationInternalBase(ClientDiagnostics clientDiagnostics, Response ra
-            await WaitForCompletionResponseAsync(DefaultPollingInterval, cancellationToken).ConfigureAwait(false);
+        public virtual async ValueTask<Response> WaitForCompletionResponseAsync(CancellationToken cancellationToken)
+        {
+            while (true)
```
Is there a way to consolidate this loop.  All of the async ones use the same implementation and all of the sync ones use the same.

archerzz replied:
I moved the logic into a separate `PollResponseAsync(pollingStrategy, cancellationToken)`.

m-nash commented on sdk/core/Azure.Core/src/Shared/OperationInternalBase.cs:
```diff
@@ -114,8 +119,21 @@ protected OperationInternalBase(ClientDiagnostics clientDiagnostics, Response ra
         /// <exception cref="RequestFailedException">Thrown if there's been any issues during the connection, or if the operation has completed with failures.</exception>
-        public virtual async ValueTask<Response> WaitForCompletionResponseAsync(CancellationToken cancellationToken) =>
-            await WaitForCompletionResponseAsync(DefaultPollingInterval, cancellationToken).ConfigureAwait(false);
+        public virtual async ValueTask<Response> WaitForCompletionResponseAsync(CancellationToken cancellationToken)
```
In the case where someone passes in a PollingInterval we need to override the set the PollingStrategy to a new instance IF its currently a LinearPollingStrategy

m-nash commented on sdk/core/Azure.Core/src/Shared/IOperationPollingStrategy.cs:
```diff
@@ -0,0 +1,19 @@
+        /// Gets interval for current iteration of polling.
+        /// </summary>
+        /// <remarks>Note that the value could change per call.</remarks>
+        public TimeSpan PollingInterval { get; }
```
I think this should be a method named GetNextWait or something like that since it will be different every time in a lot of cases.

archerzz replied:
done

m-nash commented on sdk/core/Azure.Core/src/Shared/ExponentialPollingStrategy.cs:
```diff
@@ -0,0 +1,35 @@
+            get
+            {
+                var returnValue = _pollingInterval;
+                if (_pollingInterval < maxPollingInterval)
```
I think what we want to do here is use 1 second 3 times, then double until we reach 32

so the intervals should be
1, 1, 1, 2, 4, 8, 16, 32

archerzz replied:
done

archerzz committed:
update according to review comments

1. change interface `IOperationPollingStrategy` to abstract class `OperationPollingStrategy`
  1.1 move common logic of `GetServerDelay` into `OperationPollingStrategy`
  1.2 change `PollingInterval` to `GetNextWait(response)` to be more specific and fit the new logic
2. change polling sequence of `ExponentialPollingStrategy` to `1, 1, 1, 2, 4 ... 32`
3. consolidate polling logic into `OperationInternalBase.PollingResponseAsync`
4. change implementation of `TestLroOperation.cs` to get rid of dependency on `OperationInternals` which is located in `autorest.csharp`
  4.1 I can restore the dependency by doing the similar change like `OperationInternalBase`, but that could be overkill, so I choose a simple change.
5. update test cases and build dependency

archerzz commented:
Changes after review:
1. change interface `IOperationPollingStrategy` to abstract class `OperationPollingStrategy`
  1.1 move common logic of `GetServerDelay` into `OperationPollingStrategy`
  1.2 change `PollingInterval` to `GetNextWait(response)` to be more specific and fit the new logic
2. change polling sequence of `ExponentialPollingStrategy` to `1, 1, 1, 2, 4 ... 32`
3. consolidate polling logic into `OperationInternalBase.PollingResponseAsync`
4. change implementation of `TestLroOperation.cs` to get rid of dependency on `OperationInternals` which is located in `autorest.csharp`
  4.1 I can restore the dependency by doing the similar change like `OperationInternalBase`, but that could be overkill, so I choose a simple change.
5. update test cases and build dependency

m-nash commented on sdk/core/Azure.Core/src/Shared/OperationPollingStrategy.cs:
```diff
@@ -0,0 +1,65 @@
+            return a > b ? a : b;
+        }
+
+        protected static bool TryGetRetryAfter(Response response, out TimeSpan? retryAfter)
```
I would move this into a RetryAfterPollingStrategy.  I want to separate the logic of the polling strategy from the choice of the which strategy to use.

The logic to use will do the following
If the response has retry-after use RetryAfterPollingStrategy
else if a strategy was passed in use that
else default to ConstantPollingStrategy

archerzz committed:
split polling strategy logic according to review comments

1. add two polling strategy implementation: `ZeroPollingStrategy` and `RetryAfterPollingStrategy`
  1.1 `ZeroPollingStrategy` is for testing purpose (e.g play back test recordings)
  1.2 `RetryAfterPollingStrategy` is used when detecting the server side support `retry-after` header, and corresponding logic is removed from `ConstantPollingStrategy` and `ExponentialPollingStrategy`.
2. change constructor of `OperationInternalBase` and its subclasses
  2.1 add a polling strategy parameter
  2.2 change polling strategy initialization logic, now it will"
    2.2.1 first check if there is any `retry-after` header, if so then adopt `RetryAfterPollingStrategy`
    2.2.2 then adopt the passed-in polling strategy
    2.2.3 fall back to `ConstantPollingStrategy` if none is adopted
  2.3 change the interval detemination logic of `WaitForCompletionResponseAsync(TimeSpan, CancellationToken)`. Now we adopt the max value of the passed-in interval and the interval determined by polling strategy.
3. update test cases and build dependencies

m-nash commented on sdk/core/Azure.Core/src/Shared/OperationInternalBase.cs:
```diff
@@ -14,20 +14,28 @@ namespace Azure.Core
+
+        private static OperationPollingStrategy ChoosePollingStrategy(Response rawResponse, OperationPollingStrategy? defaultPollingStrategy)
+        {
+            if (RetryAfterPollingStrategy.TryParseAndBuild(rawResponse, out RetryAfterPollingStrategy? pollingStrategy))
```
by delegating this logic we still have some coupling between the choosing and the polling strategies themselves

archerzz replied:
I realized that after we move `ChoosePollingStrategy` to `OperationPollingStrategy` abstract class, there is no coupling anymore. `TryParseAndBuild` and `TryGetRetryAfter` are put on `OperationPollingStrategy`, and it's natural for `RetryAfterPolingStrategy` to use some codes in its base class.

m-nash commented on sdk/core/Azure.Core/src/Shared/OperationInternalBase.cs:
```diff
@@ -53,10 +61,10 @@ protected OperationInternalBase(ClientDiagnostics clientDiagnostics, Response ra
+        /// Default to <see cref="ConstantPollingStrategy"/> with 1 second constant polling interval.
         /// </summary>
-        public TimeSpan DefaultPollingInterval { get; set; }
+        public OperationPollingStrategy PollingStrategy { get; set; }
```
Does this need to be public and have a setter?

m-nash commented on sdk/core/Azure.Core/src/Shared/OperationInternalBase.cs:
```diff
@@ -146,8 +166,8 @@ public virtual async ValueTask<Response> WaitForCompletionResponseAsync(TimeSpan
 
-                TimeSpan delay = GetServerDelay(response, pollingInterval);
-                await WaitAsync(delay, cancellationToken).ConfigureAwait(false);
+                TimeSpan strategyInterval = pollingStrategy.GetNextWait(response);
```
this might be better if we just instantiate a pollingstrat here since we need to call the choose logic to know if the passed in value will be used or ignored

ArcturusZhang commented on sdk/core/Azure.Core/src/Shared/RetryAfterPollingStrategy.cs:
```diff
@@ -0,0 +1,90 @@
+            return false;
+        }
+
+        protected static bool TryGetRetryAfter(Response response, out TimeSpan? retryAfter)
```
Could we add a `[MaybeNullWhen(false)]` attribute to the out parameter so that the static check will handle the null check? In this way we do not have to make the TimeSpan to be a nullable type, and every invocation of TimeSpan will not need the null-escape opeartor `!`

ArcturusZhang commented on sdk/core/Azure.Core/src/Shared/RetryAfterPollingStrategy.cs:
```diff
@@ -0,0 +1,90 @@
+        public static bool TryParseAndBuild(Response response, out RetryAfterPollingStrategy? pollingStrategy)
+        {
+            pollingStrategy = null;
+            if (TryGetRetryAfter(response, out TimeSpan? retryAfter))
+            {
+                pollingStrategy = new RetryAfterPollingStrategy(retryAfter!.Value);
+                return true;
+            }
```
For instance if we add `MaybeNullWhen` attribute, this will turn to 
```suggestion
            if (TryGetRetryAfter(response, out TimeSpan retryAfter))
            {
                pollingStrategy = new RetryAfterPollingStrategy(retryAfter.Value);
                return true;
            }
```
which is cleaner

archerzz committed:
update according to review comments

1. move `OperationInternalBase.ChoosePollingStrategy()` to `OperationPollingStrategy` class
2. make `OperationInternalBase.PollingStrategy` private and update test cases
3. update polling strategy implementation
  3.1. `ConstantPollingStrategy` will return max value of initial constant interval and the passed-in interval
  3.2. `ZeroPollingStrategy` will always return 0
  3.3. `ExponentialPollingStrategy` will always return sequence value
  3.4. `RetryAfterPollingStrategy1 will return max value of retry-after and passed-in interval
  3.5. update test cases
4. add back `OperationInternalBase.DefaultPollInterval`
5. move some common logic into `OperationPollingStrategy`

archerzz committed:
implement `OperationPoller`

1. add `OperationPoller` and `OperationPoller<T>` as the common polling logic implementation
2. replace the polling logic in `OperationInternalBase`, `Operation` and `OperationOfT`.
3. update build dependencies

m-nash committed:
wip refactor

m-nash committed:
remove unused delegate

m-nash committed:
temp - add back polling interval

m-nash committed:
add files to make autorest changes

m-nash committed:
renames to represent this is a delay strategy and separate from operations which can be used in many places

m-nash committed:
merge remote

m-nash committed:
renaming polling param

m-nash committed:
more minor refactoring

m-nash committed:
comment change

m-nash committed:
updates for non T in test framework

archerzz committed:
update test cases

archerzz committed:
fix zero interval for test recording playback

1. fix a type comparison error when using project reference (lro.BaseType.BaseType is from `Operation` in `Azure.ResourceManager`, not `Azure.Core`)
2. unify the zero interval polling logic, since we've dumped the special zero interval handling in `Operation.GetServerDelay`

m-nash committed:
tweak to code for autorest

m-nash committed:
update dataplance csproj files

m-nash committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into core/add-new-poll

m-nash committed:
missed a couple csproj files

m-nash committed:
merge main

m-nash committed:
merge from azure/main

m-nash committed:
add new files to common lro includes

m-nash committed:
update common lro files

m-nash committed:
fix folder for shared files

m-nash committed:
remove duplicate compiles

m-nash committed:
update 2 more duplicate compiles

m-nash committed:
merge from core/add-new-poll

m-nash committed:
address comments from https://github.com/Azure/autorest.csharp/pull/2000

m-nash committed:
update who owns defaulting

archerzz committed:
Update AutoRest C# version to 3.0.0-beta.20220302.4

archerzz committed:
Merge branch 'core/add-new-poll' into core/mgmt-exponential-poll

archerzz committed:
fix polling strategy tests

archerzz committed:
update latest autorest.csharp 3.0.0-beta.20220303.1

archerzz committed:
disable some operation internal tests

those tests are not working, because we no longer override `OperationInternalBase.WaitAsync`

archerzz committed:
update OperationInterceptor.cs

1. more precise reflection on the generic types
2. update `RecordedTestBaseTest`
  2.1 update `CustomOperation` due to polling architecture change
  2.2 update `WaitForCompletionDoesntWaitDuringPlayback` since in playback mode, `CustomOperation.WaitForCompletionAsync` won't be called. Instead, it's intercepted. So we need to use `DateTimeOffset` to check the interval.

m-nash committed:
update to tests

m-nash committed:
update so interceptor works with true or false scenarios for no wait in mgmt plane

m-nash committed:
merge from upstream

azure-sdk commented:
API changes have been detected in `Azure.Analytics.Synapse.Artifacts`. You can review API changes [here](https://apiview.dev/Assemblies/Review/bc586a6c89b944f28b322a791ec22376?diffOnly=True&diffRevisionId=241e6861f0764922b25034c7564eb2c2)

azure-sdk commented:
API changes have been detected in `Azure.Communication.PhoneNumbers`. You can review API changes [here](https://apiview.dev/Assemblies/Review/ad60909f89064bb48e4a716a8f4d7422?diffOnly=True&diffRevisionId=cd30fa378870482c8662975b67233f72)

**API changes**
```diff
+         public override Response<PhoneNumberSearchResult> WaitForCompletion(CancellationToken cancellationToken = default);
+         public override Response<PhoneNumberSearchResult> WaitForCompletion(TimeSpan pollingInterval, CancellationToken cancellationToken = default);
+         public override Response<PurchasedPhoneNumber> WaitForCompletion(CancellationToken cancellationToken = default);
+         public override Response<PurchasedPhoneNumber> WaitForCompletion(TimeSpan pollingInterval, CancellationToken cancellationToken = default);
```

JoshLove-msft commented on sdk/core/Azure.Core.TestFramework/src/Instrumentation/OperationInterceptor.cs:
```diff
@@ -30,46 +33,34 @@ public void Intercept(IInvocation invocation)
                 {
                     CheckArguments(invocation.Arguments);
-                    invocation.ReturnValue = InvokeWaitForCompletion(invocation.InvocationTarget, invocation.TargetType, invocation.Arguments.Last());
+                    invocation.ReturnValue = InvokeWaitForCompletion(invocation.InvocationTarget, invocation.TargetType, default);
```
Why do we no longer pass through the last argument?

JoshLove-msft commented on sdk/core/Azure.Core.TestFramework/src/Instrumentation/OperationInterceptor.cs:
```diff
@@ -30,46 +33,34 @@ public void Intercept(IInvocation invocation)
                 {
                     CheckArguments(invocation.Arguments);
-                    invocation.ReturnValue = InvokeWaitForCompletionResponse(invocation.InvocationTarget, invocation.Arguments.Last());
+                    invocation.ReturnValue = InvokeWaitForCompletionResponse(invocation.InvocationTarget as Operation, default);
```
Same question here.

archerzz committed:
fix failure in `TestResourceOperationOrResponseOfT`

Move exception throwing into `UpdateStatus` and `UpdateStatusAsync`, since the `WaitForXXX` operation will be intercepted and delegate to `OperatonPoller`.

archerzz committed:
fix test timeout in `Azure.Communication.PhoneNumbers`

1. add `InstrumentOperation` before `operation.WaitForXXX`
2. remove the zero timespan in playback mode, since the operation interceptor will poll by zero timespan

whisper6284 reviewed (commented):
Can sign off on Azure.Communication.PhoneNumbers

m-nash committed:
update api for sync wait methods

m-nash committed:
update to translation LRO to not prefecth first page

m-nash committed:
regen desktopvirtualization

annelo-msft reviewed (approved):
Synapse LGTM

m-nash committed:
add back in cancellation token forwarding

christothes commented on sdk/core/Azure.Core.TestFramework/src/Instrumentation/OperationInterceptor.cs:
```diff
@@ -16,6 +16,9 @@ internal class OperationInterceptor : IInterceptor
         internal static readonly string WaitForCompletionMethodName = nameof(Operation<object>.WaitForCompletionAsync);
         internal static readonly MethodInfo WaitForCompletionResponseAsync = typeof(Operation).GetMethod(nameof(Operation.WaitForCompletionResponseAsync), new[] { typeof(TimeSpan), typeof(CancellationToken) });
 
+        // ValueTask<Response<T>> WaitForCompletionAsync<T>(Operation<T>, TimeSpan?, CancellationToken)
```
remove?

christothes commented on sdk/core/Azure.Core/src/Shared/OperationInternalBase.cs:
```diff
@@ -17,19 +17,24 @@ internal abstract class OperationInternalBase
+            _fallbackStrategy = fallbackStrategy;
         }
 
+        //TODO: Remove after autorest changes
```
file an issue to track and mention it here?

christothes commented on sdk/core/Azure.Core/tests/TestClients/TestOperation.cs:
```diff
@@ -0,0 +1,127 @@
+
+namespace Azure.Core.Tests
+{
+    internal class TestOperation : Operation, IOperation
```
👍

christothes commented on sdk/translation/Azure.AI.Translation.Document/src/DocumentTranslationOperation.cs:
```diff
@@ -122,11 +122,6 @@ public class DocumentTranslationOperation : PageableOperation<DocumentStatusResu
         /// </summary>
         private int? _retryAfterHeaderValue;
 
-        /// <summary>
```
these should rebase out when the other PR merges, correct?

m-nash committed:
Merge branch 'main' of https://github.com/archerzz/azure-sdk-for-net into core/mgmt-exponential-poll

m-nash committed:
add pr comments

m-nash commented:
/azp run net-appconfiguration

m-nash commented:
/azp run net - appconfiguration

m-nash commented:
/azp run net - appconfiguration - ci
